[Bibi] 미션3 : 체스판 초기화 pull request 입니다.

bibi6666667 opened the pull request:
안녕하세요, 백엔드반의 Bibi입니다.
미션3 : 체스판 초기화 를 진행한 step3 브랜치 pull 요청 드립니다🙇‍♀️

<학습한 부분>
- `StringBuilder` 의 개념과 사용법 (더 학습 필요)
- 호눅스의 수업 복습 겸 다음 미션 예습 차원에서 enum과 VO에 대해 조금 학습
돌아보니 이번 미션3은 학습보다는 코딩 위주로 진행되었네요.

<진행한 내역>
- 체스판 초기화
- 체스판 출력
- 체스 게임 시작 부분 만들기 (`main()`, `start`, `end`)
- 약간의 리팩토링

지금 제 코드가 최선은 아니라고 생각하지만,
이제 1주만 남은 상황에서 우선 요구사항을 만족한 코드를 얼른 PR하고
피드백을 받으며 코드를 개선하는 게 더 낫다고 생각했습니다.
그러니 여러분의 많은 제안, 조언, 피드백, 충고 부탁드립니다. 제 학습을 도와주세요..😀

세심한 코드리뷰 늘 감사드리고, 앞으로도 잘 부탁드립니다.
늦은 시간에 알림 드려 죄송합니다.🙇‍♀️ 리뷰어 분들도 좋은 주말 되세요!

bibi6666667 committed:
[step3] initialize() - 검은색, 흰색 폰 8개 생성

우선 단계별로 구현하기 위해 요구사항 그대로
단순히 두 배열을 만들고 각 폰을 추가했었다.

bibi6666667 committed:
[step3] Pawn클래스 - representation변수 및 두 상수 추가

검은색 Pawn(P), 흰색 Pawn(p)을 나타내는 문자를
Pawn클래스의 상수로 선언하고, 이 값을 넣을 인스턴스변수인
representation을 선언했습니다.
representation을 사용해 Pawn을 만들도록 생성자를 수정했습니다.
Pawn클래스의 생성자를 사용하는 다른 클래스들도 함께
수정했습니다.

bibi6666667 committed:
[step3] Pawn클래스 - getRepresentation() 추가

PawnTest 테스트 코드를 작성하기 전에,
Pawn클래스의 인스턴스의 representation의 값을 얻을 수 있는
getRepresentation()을 추가했습니다.

bibi6666667 committed:
[step3] PawnTest - Pawn의 representation도 테스트하도록 수정

Pawn에 새로 추가된 인스턴스변수 representation이
Pawn의 인스턴스에 잘 반영되는지 확인할 수 있도록
PawnTest에 테스트 코드를 넣었습니다. (테스트 통과)

bibi6666667 committed:
[step3] Board에 whitePawns, blackPawns와 관련 메서드 추가

흰색 폰, 검은색 폰을 저장하는 List를 만들고
각 List에 넣고 확인하는 add, getSize, find메서드를 각각 만듦.

bibi6666667 committed:
[step3] Board) initialize(), pawns, size() 수정

initialize()가 지역변수가 아닌 인스턴스변수를 사용하게 하고,
중복코드를 줄임.

whitePawns와 blackPawns의 초기크기를 지정함.
실수로 삭제된 size()메소드를 다시 추가함
(BoardTest가 완성되어 필요없어지면 삭제할 예정)

bibi6666667 committed:
[step3] BoardTest에 initialize() 테스트, Board에 getPawnsResult() 추가

Board에 whitePawns, blackPawns의 결과 상태를 String으로 출력하는
getWhitePawnsResult(), getBlackPawnsResult()를 만들었습니다.

Board의 initialize()를 테스트하는 코드를 BoardTest에 작성했습니다.
(테스트 완료 - 성공)

bibi6666667 committed:
[step3] print(), getBlankLine() 추가 및 print()테스트

체스판의 빈 줄을 나타내기 위해 getBlankLine()을 만들고,
체스판을 출력하는print()를 추가했습니다.

BoardTest클래스에서 print()를 테스트했습니다.
(테스트 완료 - 성공)

bibi6666667 committed:
[step3] 검은색, 흰색 폰 각각의 메소드를 사용하도록 수정

기존 pawns가 whitePawns, blackPawns로,
add() size() find() 메소드들도 검은색, 흰색 폰 각각 따로
추가되었기 때문에 그에 맞게 테스트클래스에서
사용하는 메소드도 수정해 주었습니다.

bibi6666667 committed:
[step3] Board) 미사용 인스턴스변수와 메서드 삭제

Board클래스에서 기존 Pawns를 흰색, 검은색으로 분리하고
각각 별도의 메소드를 만듦에 따라 기존의 Pawns와
Pawns에 대한 메소드들은 사용하지 않게 되어 삭제했습니다.

bibi6666667 committed:
[step3] Board의 중복 코드 제거 및 접근제어자 변경

Board의 메서드 중 getWhitePawnsResult()와 getBlackPawnsResult()의
메서드가 많이 겹쳐, 공통 코드를 getPawnsResult()라는
private 메서드로 분리한 뒤 매개변수와 리턴값을 통해 쓰도록 했습니다.

또, addWhitePawn()의 접근제어자를 public에서 default로 변경했습니다.

bibi6666667 committed:
[step3] ChessStarter 추가) 게임 시작 및 종료

체스 게임을 시작하고 종료하는 ChessStarter 클래스를 추가했습니다.
main()이 있고 Scanner로 입력값을 받아, swith문으로 입력값에 따라
다른 결과가 나오게 했습니다.
현재는 while이 아닌 재귀를 사용해 구현한 상태입니다.

bibi6666667 committed:
[step3] ChessStarter) while을 사용해 입력값을 받도록 수정

기존 swith문과 재귀로 입력값을 반복적으로 받던 부분을
while문을 사용하면서 메소드를 분리해 호출하는 방식으로
수정했습니다.

SANGCO commented on src/main/java/chess/ChessStarter.java:
```diff
@@ -0,0 +1,46 @@
+        System.out.println("end : 게임 종료");
+    }
+
+    void ready() {
+        while (true) {
+            System.out.print("> ");
+            Scanner sc = new Scanner(System.in);
+            String input = sc.nextLine();
+            if (input.equals("start")) {
+                System.out.println("게임을 시작합니다.");
+                start();
+                break;
+            }
+            if (input.equals("end")) {
+                System.out.println("게임을 종료합니다.");
+                end();
+                break;
+            }
+            System.out.println("지정되지 않은 명령어입니다. 다시 입력해 주세요.");
+        }
+    }
```
Scanner와 같이 AutoCloseable을 구현한 클래스의 경우 
[try-with-resources](https://www.baeldung.com/java-try-with-resources)를 사용해서 쉽게 리소스를 해제할 수 있답니다.
이번 미션에서 try-with-resources 관련되어서 학습하고 적용해보면 좋을거 같아요~

SANGCO replied:
"start", "end" 이런 부분들은 상수로 만들어 사용하는건 어떨까요?

bibi6666667 replied:
네! 공유해 주신 링크 통해 공부한 뒤 적용해 보겠습니다!
start, end도 그러고 보니 상수로 선언해 쓰는 게 훨씬 낫겠네요. 개선해 보겠습니다!
코멘트 주셔서 감사합니다!🙇‍♀️

SANGCO commented on src/main/java/chess/ChessStarter.java:
```diff
@@ -0,0 +1,46 @@
+        board.print();
+    }
+
+    void end() {
+        System.exit(0);
+    }
```
너무 과한 메서드 분리가 아닐까요?

bibi6666667 replied:
ready()에 종료 명령어가 있는 게 어색해 보여서 분리했는데, 너무 짧아서 과하게 분리한 것 같네요😅 수정하겠습니다!
코멘트 감사합니다😀

SANGCO commented on src/main/java/pieces/Pawn.java:
```diff
@@ -2,18 +2,26 @@
     }
 
     public Pawn() {
-        this(WHITE);
+        this(WHITE, WHITE_REPRESENTATION);
     }
```
기본 생성자가 파라미터가 있는 생성자 보다 위로 와야 합니다.

SANGCO commented on src/main/java/pieces/Pawn.java:
```diff
@@ -2,18 +2,26 @@
 public class Pawn {
     private final String color;
+    private final String representation;
     public final static String WHITE = "white";
     public final static String BLACK = "black";
+    public final static String WHITE_REPRESENTATION = "p";
+    public final static String BLACK_REPRESENTATION = "P";
```
static이 붙은 클래스 변수가 인스턴스 변수보다 위로 가야 합니다.

bibi6666667 replied:
그렇군요! 기본 생성자, 클래스변수 위치 모두 수정하겠습니다.
코멘트 감사합니다🙇‍♀️

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -6,18 +6,79 @@
-        pawns.add(pawn);
+    void addWhitePawn(Pawn pawn) {
+        whitePawns.add(pawn);
     }
 
-    public int size() {
-        return pawns.size();
+    void addBlackPawn(Pawn pawn) {
+        blackPawns.add(pawn);
```
흰색 Pawn을 추가하는 메서드에 검은색 Pawn을 넣는 테스트 코드를 만들어 봅시다.

bibi6666667 replied:
아무 이상 없이 통과가 되네요😅 요구사항만 따라가느라 Pawn을 확인하는 코드를 짜지 않았네요. Pawn의 색을 확인하고 추가하도록 수정하고, 테스트도 해보겠습니다!
코멘트 감사합니다🙇‍♀️!

SANGCO commented on src/test/java/chess/BoardTest.java:
```diff
@@ -39,4 +39,19 @@ void putOther() {
+        board.initialize();
+        board.print();
+    }
 }
```
![image](https://user-images.githubusercontent.com/26179444/107593824-b59d4380-6c53-11eb-96a2-8ce5bd465118.png)
IntelliJ 설정을 찾아서 적용해 보아요.
[How to make IntelliJ IDEA insert a new line at every end of file?](https://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file)
[파일 끝에 개행을 추가해야 하는 이유](https://blog.coderifleman.com/2015/04/04/text-files-end-with-a-newline/)

bibi6666667 replied:
앗 저번에 공부했던 부분인데 신경쓰지 못했네요.. 죄송합니다😭
개행 추가하고 인텔리제이 설정도 따라서 해 놓겠습니다!
코멘트도 주시고 링크 공유해주셔서 감사합니다😀👍

SANGCO reviewed (changes requested):
Bibi님 구현 잘하셨네요. 고생하셨습니다!
리뷰 남겼으니 확인해 보시고 질문 있으시면 댓글 부탁드립니다~

tmdgusya commented on src/main/java/chess/ChessStarter.java:
```diff
@@ -6,38 +6,41 @@
+    void ready() {
+        while (true) {
+            System.out.print("> ");
+            Scanner sc = new Scanner(System.in);
```
스캐너가 반복문안에 있어야 할까요?

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -20,4 +20,18 @@ public Pawn findPawn(int index) {
+        Pawn blackPawn = new Pawn(Pawn.BLACK);
+        List<Pawn> blackPawns = new ArrayList<>(8);
+        for (int i = 0; i < 8; i++) {
+            blackPawns.add(blackPawn);
```
공통적으로 8이라는 숫자가 같은 의미로 사용되는데
해당 8을 상수로빼서 MAX_SIZE 라던가 BOARD_SIZE 등의 상수로 추출해내는 것은 어떨까요?

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -7,19 +7,45 @@
+
     private void initialize() {
         Pawn whitePawn = new Pawn();
         List<Pawn> whitePawns = new ArrayList<>(8);
```
생각해보니 10~11 Line 에서 new ArrayList() 가 이루어지는데 아래에서 재할당할 이유가 있을까요?

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -48,13 +48,9 @@ public Pawn findBlackPawn(int index) {
-        List<Pawn> blackPawns = new ArrayList<>(8);
         for (int i = 0; i < 8; i++) {
+            whitePawns.add(whitePawn);
             blackPawns.add(blackPawn);
```
이경우 현재 동일 인스턴스를 add 하는 것으로 보여지는데 어떤 문제가 일어날 수 있을까요.
사실 지금 상태에서는 충분히 문제를 일으키지 않을 것 같은데,
후에 체스에 대한 조작이 일어난다면, 체스 객체자체가 움직이는 것이 아니라 보드가 체스를 움직이는 느낌이 들것 같습니다.
해당 방식에서는 Pawn 이 이동시 blackpawn 하나만 이동해도 같은 객체라 모두 이동하게 될것으로 보이기 때문입니다.

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -34,6 +34,22 @@ public int getBlackPawnsSize() {
 
+    public String getWhitePawnsResult(){
+        StringBuilder sb = new StringBuilder();
+        for (Pawn aPawn : whitePawns) {
```
aPawn 이라는 Naming 은 적절하지 않은것 같습니다.. 어떤 의도인지 잘 모르겠네요.
차라리 pawn 으로하는것이 낫지 않았을까요?

tmdgusya commented on src/test/java/chess/BoardTest.java:
```diff
@@ -19,18 +19,18 @@ void setBoard() {
     @DisplayName("체스판에 흰색 폰을 추가 후 갯수와 목록을 확인한다.")
     void putWhitePawn() {
```
putWhitePawn -> addWhitePawn 가 지금 테스트 케이스 제목에 적절하지 않을까라는 생각을 해봅니다.
어디까지나 개인적인 견해입니다.

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -6,14 +6,9 @@
 public class Board {
-    private List<Pawn> pawns = new ArrayList<>();
     private List<Pawn> whitePawns = new ArrayList<>(8);
     private List<Pawn> blackPawns = new ArrayList<>(8);
```
위에서 언급한것과 마찬가지로 이부분에서도 8이 이용되고 있는데
상수로 추출하는것이 적절해보입니다.

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -9,11 +9,11 @@
     private List<Pawn> blackPawns = new ArrayList<>(8);
 
-    public void addWhitePawn(Pawn pawn) {
+    void addWhitePawn(Pawn pawn) {
```
접근제어자가 필요해 보입니다.

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -9,11 +9,11 @@
     }
 
-    public void addBlackPawn(Pawn pawn) {
+    void addBlackPawn(Pawn pawn) {
```
접근제어자가 필요해 보입니다.

tmdgusya reviewed (commented):
비비 제가 그룹 코드리뷰때 상세하게 봐드리지 못한것 같아서
이렇게 리뷰남기러 왔습니다! 주말엔 푹 쉬시고 제 Comment 부분도 한번 생각해보시고 
답글 달아주시면 감사할듯 합니다 👍

bibi6666667 committed:
[step3] 피드백) START, END 상수 추가

start, end를 String 대신 상수 지역변수로 만들어 사용했습니다.

bibi6666667 committed:
[step3]피드백) end() 삭제

end()메서드를 삭제사고, end()를 분리하기 전 상태로 ready()를
수정했습니다.

bibi6666667 committed:
[step3]피드백) static변수, 기본 생성자 위치 변경

static변수의 위치를 인스턴스 변수 윗 줄로 옮겼습니다.
기본 생성자를 다른 생성자보다 위에 위치하도록 했습니다.

bibi6666667 committed:
[step3]피드백)addWhitePawn()에 검은 폰을 넣는 테스트 추가

addWhitePawn()에 검은 폰을 넣는 테스트 코드를 추가했습니다.
폰을 확인하지 않고 추가하는 메서드라 테스트가 통과되었습니다.
각 메서드에서 Pawn을 확인하고 추가하도록 메서드를
수정하겠습니다.

bibi6666667 committed:
[step3]피드백) addWhitePawn(), addBlackPawn() 수정 및 테스트 (성공)

Board클래스의 addWhitePawn(), addBlackPawn()에 폰의 색깔을
확인하는 코드를 추가했습니다.

BoardTest에 addWhitePawn()의 테스트코드를 수정했습니다.
(테스트 완료 - 성공)

bibi6666667 committed:
[step3] Scanner를 반복문 밖에서 생성하도록 수정

bibi6666667 committed:
[step3] Board에 값이 8인 PAWN_NUM 상수 추가 및 관련 코드 수정

8이라는 폰의 숫자를 PAWN_NUM이라는 상수로 선언한 뒤,
8을 사용하던 ArrayList와 메서드들을 상수를 사용하도록
수정했습니다.

bibi6666667 committed:
[step3] 피드백) try-with-resources로 스캐너 닫기

링크 주신 try-with-resources에 대해 공부하고, ChessStarter에서
스캐너를 사용하는 부분에 적용해 보았습니다.

bibi6666667 commented:
> 비비 제가 그룹 코드리뷰때 상세하게 봐드리지 못한것 같아서
> 이렇게 리뷰남기러 왔습니다! 주말엔 푹 쉬시고 제 Comment 부분도 한번 생각해보시고
> 답글 달아주시면 감사할듯 합니다 👍

@tmdgusya 코드리뷰 감사드립니다 로치!😄
코멘트 주신 부분들 수정하고 생각해 봤습니다.
(로치가 남겨 주신 코멘트가 이메일에서는 보이는데 이 PR에는 코멘트 창이 안 보여서..🤔 부득이하게 여기에 대신 정리해서 남깁니다.)

<수정한 부분>
- `Scanner`를 반복문 밖에 선언했습니다.
- `8`이라는 숫자도 `PAWN_NUM`이라는 이름의 상수로 사용했습니다.

<아직 수정하지 않은 부분>
- `Board`클래스의 `addWhitePawn()`, `addBlackPawn()` 의 접근제어자
  - 저는 Board에 임의로 외부에서 폰을 추가할 수 없게끔 일부러 `public`을 빼고, 테스트코드에서는 사용해야 하니까 `private`는 붙이지 않고 default로 두었습니다. 더 나은 방법이 있다면 개선하고 싶은데, 어떤 방법이 좋은지 추천해 주실 수 있나요?🤔
- 메서드명, 변수명 관련 조언
  - 이 부분은 다음 스텝 진행하면서 한 번 바꿔 보겠습니다! 
- 동일 인스턴스를 add하고 있다는 조언
  - 혹시 `whitePawns.add(whitePawn)`, `blackPawns.add(blackPawn)`부분을 말씀하시는 게 맞나요? 제 코드에서 `whitePawns`는 ArrayList, `whitePawn`은 흰색으로 선언된 `Pawn`이라 동일 인스턴스가 더해지는 상태는 아니라고 생각했습니다.
- Board.java 에서 `new ArrayList`를 재할당하고 있다는 조언
  - 제 코드에서는 예전에 수정된 부분인데 왜 PR에서 이렇게 나타날까요?😅 그래도 코멘트 감사합니다!

신경 써서 피드백과 조언 주셔서 감사합니다 로치!
덕분에 제 코드가 많이 개선되었네요😀👍

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -6,18 +6,88 @@
 
-    public void add(Pawn pawn) {
-        pawns.add(pawn);
+    boolean addWhitePawn(Pawn pawn) {
```
접근 제한자를 왜 디폴트로 하셨는지 궁금합니다~

bibi6666667 replied:
안녕하세요 상코! 추가 리뷰 주셔서 감사합니다😀👍

제 의도는 Board에 임의로 외부에서 폰을 추가할 수 없게끔 하는 것이었습니다.
그래서 public을 뺐지만, 테스트코드에서는 해당 메소드를 사용해야 하니 private는 붙이지 않았습니다. 
더 나은 방법이 있다면 개선하고 싶은데, 어떤 접근제어자가 적당할지 잘 몰라서 default로 두었습니다.
 (접근제한자에 대해 공부한 적은 있지만 사용해 본 적은 별로 없어서.. 조언 부탁드립니다😅)

SANGCO replied:
접근 제한자를 선택하실 때 우선 public 혹은 private 만들고
private인데 상속 관계에 있는 객체들에서 사용해야 한다고 하면 protected로
private 인데 다른 패키지에서 사용을 해야 한다면 package-private(default)로
이런식으로 접근을 하면 될거 같습니다.

위에 테스트를 위해서 디폴트로 접근 제한자를 변경 하신건 잘하셨구요.
private 메소드를 테스트 하는 방법이 있는데
그거는 아마 뒤에 나오지 싶습니다.
우선은 이렇게 하셔도 괜찮을거 같아요~

bibi6666667 replied:
오.. 그렇군요! 적어 두고 다음 미션 때도 이런 부분들 생각하면서 접근제어자를 사용해 보겠습니다.
상세히 답변 주셔서 감사합니다!🙇‍♀️👍

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -6,18 +6,88 @@
 
-    public int size() {
-        return pawns.size();
+    boolean addBlackPawn(Pawn pawn) {
```
여기도 접근제한자

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -6,18 +6,88 @@
+    }
+
+    void initialize() {
+        Pawn whitePawn = new Pawn();
+        Pawn blackPawn = new Pawn(Pawn.BLACK, Pawn.BLACK_REPRESENTATION);
+        for (int i = 0; i < PAWN_NUM; i++) {
+            whitePawns.add(whitePawn);
+            blackPawns.add(blackPawn);
+        }
```
```java
Pawn whitePawn = new Pawn();
Pawn blackPawn = new Pawn(Pawn.BLACK, Pawn.BLACK_REPRESENTATION);
```
이 코드가 for문 밖에 있는것과 안에 있는게 어떤 차이가 있을까요?

그리고 위에 만든 addWhitePawn(), addBlackPawn() 메서드가 
여기서 쓰려고 만드신게 아닌가요?

bibi6666667 replied:
for문 안에 있으면 매번 새로운 Pawn객체를 만들고 그 객체를 추가시키는 반면,
for문 밖에 있으면 만든 객체 하나를 가지고 반복해서 추가하는 차이가 있다고 생각합니다.
addWhitePawn(), addBlackPawn()은 그냥 요구사항대로 만든 메소드인데, 듣고 보니 initialize()에서 사용할 수 있네요. 만들어 놓고 왜 사용을 안 한 건지...😅 수정하겠습니다!
추가 리뷰 주셔서 감사합니다🙇‍♀️🙇‍♀️

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -6,18 +6,88 @@
+        return blackPawns.get(index);
+    }
+
+    void initialize() {
```
여기도 접근제한자가 디폴트이네요.
참고로 private 메서드는 보통 private 메서드를 사용하고 있는 메서드 밑에 위치 시킵니다.

bibi6666667 replied:
그렇군요! private메서드들의 위치 수정하겠습니다. 감사합니다🙇‍♀️

SANGCO reviewed (changes requested):
추가로 간단한 리뷰 남깁니다~

bibi6666667 committed:
[step3] 피드백) private메서드들의 위치를 수정

모든 private메서드들의 위치를 해당 메서드를 사용하는 메서드
아래로 옮겼습니다.

bibi6666667 committed:
[step3] 피드백) Board의 initialize() 수정

initialize()에서 기존에 만들어 둔 메서드들을 사용하도록 코드를
수정했습니다.

SANGCO reviewed (approved):
수정 잘하셨네요. 고생 많으셨습니다~

```java
Pawn whitePawn = new Pawn();
Pawn blackPawn = new Pawn(Pawn.BLACK, Pawn.BLACK_REPRESENTATION);
```

이 부분 리뷰 관련해서는 이번 미션과 다음 미션 까지도 이렇게 가도 상관은 없는데
아마 딱 걸리는 미션이 있을 겁니다.
거기서 뭐 수정해도 좋을거 같네요.
다음 미션으로 넘어 가시죠~
